source/ceylon/collection/ReversibleHashMap.ceylon

```diff
@@ -0,0 +1,110 @@
+"A [[Map]] which may be [[reversed|reverse]], backed by two
+ [[HashMap]]s. The reversible mapping from [[Key]]s to 
+ [[Item]]s is a _bijection_ (a one-to-one mapping).
+ 
+ In order to maintain the bijection when the 
+ `ReversibleHashMap` is mutated, the [[put]] operation not 
+ only overwrites any existing mapping for the given key, but
+ also removes any existing entry with the given item. Note
+ that this behavior, in principle, violates the general
+ contract of [[MutableMap.put]]."
+//TODO: generalize this implementation to also allow
+//      ReversibleTreeMaps!
+shared class ReversibleHashMap<Key, Item>
+        (stability = linked, hashtable = Hashtable(),
+         entries = {}) 
+        satisfies MutableMap<Key, Item> 
+        given Key satisfies Object
+        given Item satisfies Object {
+    
+    "Performance-related settings for the backing arrays."
+    Hashtable hashtable;
+    
+    "Determines whether this is a linked hash map with a
+     stable iteration order."
+    Stability stability;
+    
+    "The initial entries in the map."
+    {<Key->Item>*} entries;
+    
+    variable MutableMap<Key, Item> map 
+            = HashMap<Key, Item>
+                    (stability, hashtable, entries);
+    variable MutableMap<Item, Key> reverseMap 
+            = HashMap<Item, Key>
+                    (stability, hashtable);
+    for (key->item in entries) {
+        reverseMap.put(item, key);
+    }
+    
```

Comment:
Use `put(key,item)` instead for initialization in case of duplicate keys or items or non-repeatable iteratables?

*Possibly* throw for duplicate items here and for `putAll`? FWIW, Guava's `BiMap` throws in these cases and also on `put`, and offers `forcePut`. On the other hand, no one throws on pre-existing *`key`s*, so I guess that's a bit inconsistent.